src/network/network_gui.cpp

```diff
@@ -1761,166 +1743,538 @@ static void PopupClientList(ClientID client_id, int x, int y)
+			if (c != nullptr && ci->client_playas != c->index) continue;
+			if (c == nullptr && ci->client_playas != COMPANY_SPECTATOR) continue;
+			if (StrEmpty(ci->client_name)) continue;
+			has_players = true;
+
+			/* Draw the player line (if in range of scrollbar). */
+			if (count >= count_start && count < count_end) {
+				uint x = rtl ? text_left : text_right;
+
+				/* If there are buttons for this client, draw them. */
+				auto button_find = this->buttons.find(count);
+				if (button_find != this->buttons.end()) {
+					x = this->DrawButtons(x, y, &button_find->second);
+				}
+
+				StringID client_string = STR_JUST_RAW_STRING;
+
+				if (ci->client_id == CLIENT_ID_SERVER) {
+					client_string = STR_NETWORK_CLIENT_LIST_PLAYER_HOST;
+				}
+				if (ci->client_id == _network_own_client_id) {
+					client_string = STR_NETWORK_CLIENT_LIST_PLAYER_SELF;
+				}
+
+				SetDParamStr(0, ci->client_name);
+				DrawString(rtl ? x : text_left + CLIENT_OFFSET_LEFT, rtl ? text_right - CLIENT_OFFSET_LEFT : x, y + text_y_offset, client_string, TC_BLACK);
+			}
+
+			y += this->line_height;
+			count++;
 		}
 
-		/* Find the new selected item (if any) */
-		pt.y -= this->GetWidget<NWidgetBase>(WID_CL_PANEL)->pos_y;
-		int item = -1;
-		if (IsInsideMM(pt.y, WD_FRAMERECT_TOP, this->GetWidget<NWidgetBase>(WID_CL_PANEL)->current_y - WD_FRAMERECT_BOTTOM)) {
-			item = (pt.y - WD_FRAMERECT_TOP) / this->line_height;
+		/* Mark companies without a player more clearly. */
+		if (!has_players) {
+			if (count >= count_start && count < count_end) {
+				if (c != nullptr) {
+					DrawString(rtl ? text_left : text_left + CLIENT_OFFSET_LEFT, rtl ? text_right - CLIENT_OFFSET_LEFT : text_right, y + text_y_offset, STR_NETWORK_CLIENT_LIST_NO_PLAYERS, TC_WHITE);
+				} else {
+					DrawString(rtl ? text_left : text_left + CLIENT_OFFSET_LEFT, rtl ? text_right - CLIENT_OFFSET_LEFT : text_right, y + text_y_offset, STR_NETWORK_CLIENT_LIST_NO_SPECTATORS, TC_WHITE);
+				}
+			}
+
+			y += this->line_height;
+			count++;
 		}
 
-		/* It did not change.. no update! */
-		if (item == this->selected_item) return;
-		this->selected_item = item;
+		return count;
+	}
 
-		/* Repaint */
-		this->SetDirty();
+	void DrawWidget(const Rect &r, int widget) const override
+	{
+		switch (widget) {
+			case WID_CL_SERVER_NAME: {
+				SetDParamStr(0, _settings_client.network.server_name);
+				int offset = std::max(0, ((int)(r.bottom - r.top + 1) - (int)FONT_HEIGHT_NORMAL) / 2); // Offset for rendering the text vertically centered
+				DrawString(r.left, r.right, r.top + offset, STR_BLACK_RAW_STRING, TC_BLACK, SA_RIGHT);
+			} break;
+
+			case WID_CL_CLIENT_NAME: {
+				SetDParamStr(0, _settings_client.network.client_name);
+				int offset = std::max(0, ((int)(r.bottom - r.top + 1) - (int)FONT_HEIGHT_NORMAL) / 2); // Offset for rendering the text vertically centered
+				DrawString(r.left, r.right, r.top + offset, STR_BLACK_RAW_STRING, TC_BLACK, SA_RIGHT);
+			} break;
```

Comment:
You can use the existing CenterBounds() function on these two, or just wait for #9063 and not have to do the manual drawing at all.